### PR TITLE
Add librdkafka based consumer

### PR DESCRIPTION
**Pull request by tkornai**
Hence solving for #350

Only tested locally, not ready for production use yet.

Checking the last commit in this PR shows diffs agains the in_kafka_group implementation which I used as a template.

### CODE REVIEW

**Comment by repeatedly** [comment]
I have just commented on old PR. Check my review: https://github.com/fluent/fluent-plugin-kafka/pull/366#pullrequestreview-483880657

File: lib/fluent/plugin/in_rdkafka_group.rb [hunk 1]
```diff
@@ -236,10 +174,42 @@ def reconnect_consumer
...
+    attr_reader :topic
+    attr_reader :messages
+
+    def initialize(topic)
+      @topic = topic
+      @messages = []
+    end
+  end
+
+  def each_batch(&block)
```
Comment (robinp-tw) [inline lib/fluent/plugin/in_rdkafka_group.rb:187]: Please add some comments on the purpose of new classes / methods where non-trivial.

**Review by robinp-tw — COMMENTED** [review]
Random comments.

File: lib/fluent/plugin/in_rdkafka_group.rb [hunk 2]
```diff
@@ -193,25 +141,15 @@ def shutdown
     # If the problem happens, we will add a guard for consumer.
     consumer = @consumer
     @consumer = nil
-    consumer.stop
+    consumer.close
```
[the remark this answers is not in the document]
Comment (tkornai) [reply lib/fluent/plugin/in_rdkafka_group.rb:135]: I couldn't find anything about thread safetiness neither in the two below sources:
- [rdkafka-ruby](https://github.com/appsignal/rdkafka-ruby) 
- [librdkafka](https://github.com/edenhill/librdkafka)

Do you have anything specific in mind regarding what could go wrong?

[the remark this answers is not in the document]
Comment (tkornai) [reply lib/fluent/plugin/in_rdkafka_group.rb:135]: Ah, actually I found it: https://github.com/edenhill/librdkafka/wiki/FAQ#is-the-library-thread-safe
I think we are safe.

File: lib/fluent/plugin/in_rdkafka_group.rb [hunk 3]
```diff
@@ -193,25 +141,15 @@ def shutdown
...
-      if m = /^\/(.+)\/$/.match(topic)
-        topic_or_regex = Regexp.new(m[1])
-        $log.info "Subscribe to topics matching the regex #{topic}"
-      else
-        topic_or_regex = topic
-        $log.info "Subscribe to topic #{topic}"
-      end
-      consumer.subscribe(topic_or_regex, start_from_beginning: @start_from_beginning, max_bytes_per_partition: @max_bytes)
-    }
+    consumer = Rdkafka::Config.new(@kafka_configs).consumer
```
[the remark this answers is not in the document]
Comment (tkornai) [reply lib/fluent/plugin/in_rdkafka_group.rb:141]: rdkafka is also just a pass-through to librdkafka: https://github.com/appsignal/rdkafka-ruby/blob/01adcce4565631cecede5c7401a42b9064ad72c3/lib/rdkafka/config.rb#L4-L6

[the remark this answers is not in the document]
Comment (tkornai) [reply lib/fluent/plugin/in_rdkafka_group.rb:141]: Also this is sort of a requirement from librdkafka: https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#expose-the-configuration-interface-pass-thru

File: lib/fluent/plugin/in_rdkafka_group.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (robinp-tw) [reply lib/fluent/plugin/in_rdkafka_group.rb:135]: Sounds good, thanks!

File: lib/fluent/plugin/in_rdkafka_group.rb (hunk 1 above)
Comment (tkornai) [reply lib/fluent/plugin/in_rdkafka_group.rb:182]: Good point, added some comments.

**Comment by tkornai** [comment]
Thanks @repeatedly and @robinp-tw for the comments, I tried to address all of them.

File: lib/fluent/plugin/in_rdkafka_group.rb [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+  # advances thes consumer offset, so in case the process crashes we might lose at most :max_batch_size messages.
+  def each_batch(&block)
+    batch = nil
+    message = nil
+    while @consumer
+      message = @consumer.poll(@max_wait_time_ms)
+      if message
+        if not batch
+          batch = Batch.new(message.topic)
+        elsif batch.topic != message.topic || batch.messages.count >= @max_batch_size
```
Comment (repeatedly) [inline lib/fluent/plugin/in_rdkafka_group.rb:190]: Use `messages.size` instead of `messages.count`.

**Comment by tkornai** [comment]
Hi @repeatedly, thanks for your help with this change set.

Quick question regarding the next steps. Should I add a disclaimer somewhere about the new input plugin being in a "beta" state? Or is there anything else that I should do so the PR can be merged?

**Comment by repeatedly** [comment]
> Should I add a disclaimer somewhere about the new input plugin being in a "beta" state? 

If you want to put it, there are 2 approaches:

- Use log.warn in `configure` method
- Write it on README section

**Comment by tkornai** [comment]
Decided to do both. Is that anything else I should take care of?

**Comment by repeatedly** [comment]
Merged. I will release new version soon.

**Pull request merged**